### ISSUE TITLE
Feature Request: Run migrations outside of Wekan, before starting Wekan

### ISSUE THREAD

**Issue by xet7** [wekan/wekan#1632]
## Issue

**Problem description**
* There is 20 instances running Wekan
* New Wekan version is changed to production
* Every instance starts migrations at once
* There is no locking
* Instances mess up migrations

**TODO**
* Move all migrations to be separate from Wekan.
* Add debug logs to migrations.
* When new Wekan release comes, check and coordinate that all instances goes to service break mode
* Then only one instance runs migration code
* When migration is finished, coordinate that all instances start running new Wekan version

**Comment by xet7** [comment]
One way to do this manually, is that when updating:

1) Shutdown other Wekan Docker containers, and only leave one Wekan Docker container running.

2) Then do the update, let migrations finish.

3) Start other containers.

**Comment by xet7** [comment]
Currently in progress of removing all migrations, so it works with any existing data, and WeKan starts immediately.

**Comment by mfilser** [comment]
@xet7 
I think Migrations in general should still keep. Every Software needs Migration. Having no migrations will result in more complicated code, and more database queries, which will result in a slower app.

Best would be the first instance makes the migrations and set a lock, other instances will wait until Migration completes at first instance.

**Comment by xet7** [comment]
@mfilser 

Migrations are many hours of downtime for big databases that have size of 500 GB or 800 GB. I can not release WeKan version that causes global downtime of many hours for about 10k servers.

**Comment by xet7** [comment]
@mfilser 

Also, often there is no enough free disk space for migrations.

**Comment by mfilser** [comment]
@xet7
This are huge databases, but I think most databases aren't big like this.
Also,  I think databases with this size have attachments into it Attachments should be stored outside of the database. The attachment migration has maybe a longer downtime, but could be planned by the admin. And if other migrations need more than a few minutes, that migrations arent't developed well.

And at the end, I keep it, having migrations in Wekan is a "must have" in Wekan and other softwares

**Comment by xet7** [comment]
@mfilser 

> This are huge databases, but I think most databases aren't big like this.

But I have these problems, migrating big databases, losing many hours of time.

> Also, I think databases with this size have attachments into it Attachments should be stored outside of the database. 

Yes, I'm adding this feature.

> The attachment migration has maybe a longer downtime, but could be planned by the admin. 

No, there will be no downtime. Migration settings will be at Admin Panel, and happen at background.

> And if other migrations need more than a few minutes, that migrations arent't developed well.

Yes, current wekan/server/migrations.js takes many hours to run for big databases, they are not developed well.

> And at the end, I keep it, having migrations in Wekan is a "must have" and other softwares

I'm developing these so that it will be possible to upgrade from any old WeKan version to newest, with no downtime. I do not know why you insist on having downtime.

**Comment by xet7** [comment]
@mfilser 

In the future, migrations of files and databases will be optional, and started at Admin Panel / Settings / Migrations. They will run at background, without requiring downtime.

Any file and text data can be at any file storage (MongoDB, filesystem, S3 etc) or any database (any version of MongoDB, SQLite, etc).

When data is moved at background, and there is new request for file or database query result, if it is not at source, it will get it from target.

**Comment by mfilser** [comment]
Are this features planed for Meteor Wekan?
Migration of files from database to filesystem was already done by me before 2 years.

**Comment by xet7** [comment]
@mfilser 

> Are this features planed for Meteor Wekan?

Yes, Meteor WeKan will have many internal changes, changing it to custom framework, that supports many databases etc. During changes, all features stay working, bugs are fixed, etc.

> Migration of files from database to filesystem was already done by me before 2 years.

I think currently WeKan always uses files from MongoDB. Not filesystem.

**Comment by mfilser** [comment]
@xet7 
> Yes, Meteor WeKan will have many internal changes, changing it to custom framework, that supports many databases etc. During changes, all features stay working, bugs are fixed, etc.

What's your progress of developing this ?

**Comment by xet7** [comment]
@mfilser 

I'm currently developing this, to get all WeKan platforms upgraded to newest WeKan.

**Comment by mfilser** [comment]
Do you have a time scheduling for this development?

**Comment by xet7** [comment]
I try to get all platforms up-to-date and all paid features implemented during this year.
